File: output-parser-test-py/src/stream_tool_calls_parser.py
"""stream_tool_calls_parser.py - 使用工具解析器解析流式工具结果"""
from langchain_core.output_parsers.openai_tools import JsonOutputToolsParser
from pydantic import BaseModel, Field

from utils import create_chat_model


class ScientistInfo(BaseModel):
    name: str = Field(description="科学家的全名")
    birth_year: int = Field(description="出生年份")
    death_year: int | None = Field(default=None, description="去世年份")
    nationality: str = Field(description="国籍")
    fields: list[str] = Field(description="研究领域列表")
    achievements: list[str] = Field(description="主要成就")
    biography: str = Field(description="简短传记")


model = create_chat_model()
model_with_tool = model.bind_tools([ScientistInfo])
parser = JsonOutputToolsParser()
chain = model_with_tool | parser


async def main() -> None:
    try:
        stream = chain.astream("详细介绍牛顿的生平和成就")
        last_content = ""
        print("📡 实时输出流式内容:\n")
        async for chunk in stream:
            if chunk:
                tool_call = chunk[0]
                current_content = str(tool_call.get("args", {}))
                if len(current_content) > len(last_content):
                    print(current_content[len(last_content):], end="")
                    last_content = current_content
        print("\n\n✅ 流式输出完成")
    except Exception as error:
        print(f"\n❌ 错误: {error}")


if __name__ == "__main__":
    import asyncio

    asyncio.run(main())

File: output-parser-test/src/stream-tool-calls-parser.mjs
import 'dotenv/config';
import { ChatOpenAI } from '@langchain/openai';
import { JsonOutputToolsParser } from '@langchain/core/output_parsers/openai_tools';
import { z } from 'zod';

const model = new ChatOpenAI({
    modelName: process.env.MODEL_NAME,
    apiKey: process.env.OPENAI_API_KEY,
    temperature: 0,
    configuration: {
        baseURL: process.env.OPENAI_BASE_URL,
    },
});

// 定义结构化输出的 schema
const scientistSchema = z.object({
    name: z.string().describe("科学家的全名"),
    birth_year: z.number().describe("出生年份"),
    death_year: z.number().optional().describe("去世年份，如果还在世则不填"),
    nationality: z.string().describe("国籍"),
    fields: z.array(z.string()).describe("研究领域列表"),
    achievements: z.array(z.string()).describe("主要成就"),
    biography: z.string().describe("简短传记")
});

// 绑定工具到模型
const modelWithTool = model.bindTools([
    {
        name: "extract_scientist_info",
        description: "提取和结构化科学家的详细信息",
        schema: scientistSchema
    }
]);

// 1. 绑定工具并挂载解析器
const parser = new JsonOutputToolsParser();
const chain = modelWithTool.pipe(parser);

try {
    // 2. 开启流
    const stream = await chain.stream("详细介绍牛顿的生平和成就");

    let lastContent = ""; // 记录已打印的完整内容
    let finalResult = null; // 存储最终的完整结果

    console.log("📡 实时输出流式内容:\n");

    for await (const chunk of stream) {
        // console.log(chunk);

        if (chunk.length > 0) {
            const toolCall = chunk[0];

            // 获取当前工具调用的完整参数内容
            const currentContent = JSON.stringify(toolCall.args || {}, null, 2);

            if (currentContent.length > lastContent.length) {
                const newText = currentContent.slice(lastContent.length);
                process.stdout.write(newText); // 实时输出到控制台
                lastContent = currentContent; // 更新已读进度
            }

            console.log(toolCall.args);
        }
    }

    console.log("\n\n✅ 流式输出完成");

} catch (error) {
    console.error("\n❌ 错误:", error.message);
    console.error(error);
}